android/src/app/organicmaps/car/screens/hacks/PopToRootHack.java

```diff
@@ -0,0 +1,116 @@
+package app.organicmaps.car.screens.hacks;
+
+import static java.util.Objects.requireNonNull;
+
+import android.os.Handler;
+import android.os.Looper;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.car.app.CarContext;
+import androidx.car.app.model.Template;
+import androidx.car.app.navigation.model.RoutePreviewNavigationTemplate;
+import androidx.lifecycle.LifecycleOwner;
+
+import app.organicmaps.car.screens.base.BaseScreen;
+
+/**
+ * This class is used to provide a working solution for next the two actions:
+ * <ul>
+ *   <li> {@code popToRoot();}
+ *   <li> {@code push(new Screen());}
+ * </ul>
```

Comment:
I am not sure that I've got the problem. Could you please provide an example of workflow that doesn't work without this "hack"?

Reply:
It's really hard to explain but I'll try.

The main problem is that we're limited to 6 (or 5, I don't remember) screens in the stack. The 6th(5th) one can only be a navigation template. So, we can create only 5 screens. Looks like it should be enough and we'll never need more. But the thing is that one screen can be treated as 2 or more during the information updates on it. It depends on the template and on the information being updated. The description can be checked on the docs page for each template. Example for `MapTemplate`: https://developer.android.com/reference/androidx/car/app/navigation/model/MapTemplate

>Template Restrictions In regards to template refreshes, as described in [onGetTemplate](https://developer.android.com/reference/androidx/car/app/Screen#onGetTemplate()), this template is considered a refresh of a previous one if:
>* The template title has not changed, and the number of rows and the title (not counting spans) of each row between the previous and new [Pane](https://developer.android.com/reference/androidx/car/app/model/Pane)s or [ItemList](https://developer.android.com/reference/androidx/car/app/model/ItemList)s have not changed.

This `MapTemplate` is used on `PlaceScreen`.
So, imagine the workflow:
* You are on the main screen (screen stack size: 1)
* You go to search (screen stack size: 2)
* You got then to interactive search (screen stack size: 3)
* You go to place screen (screen stack size: 4)
* You click the "build route" button. The route is built and the screen is refreshed (added a new row with the info about route length and time). According to the template restrictions, this refresh will be considered as a new screen. (screen stack size: 5) We already reached the limit.
* What if we want to change the routing settings? It's not good to drop all the progress (all these screens), go to the main screen, then to settings and so on. So, we have a settings button on this screen. If we click - we'll have 6 screens in the stack. That's not possible.
* Another case. We support multiple routes. And one more screen will be to select the track. It's also not possible.
* Or one more case. We want to go to another place and select it on the map. It also will be considered as a new screen, so 6 screens in the stack again.

That's why we have to pop all previous screens when we push the new `PlaceScreen`. That means when we go back from this screen, we'll pop directly to the main screen. All search screens will be lost. That's not very convenient but that's the only way.

And now to the question "Why do we need this `PopToRootHack` and why can't we just call `ScreenManager.popToRoot()`?"

Imagine the case:
You're on a search screen and you select the place. What should be called?
```java
ScreenManager.popToRoot();
ScreenManager.push(new PlaceScreen(...));
```

What should happen? The screens stack should be cleaned and a new screen should be pushed to the stack. At the end, the stack size should be equal to 2: root screen and place screen.
And that's how it works 😆 
But there is one problem. 
There are TWO stacks: one for the screen objects (and it's size will be 2, as I wrote above) and another one (let's call it the callstack). The second stack represents the screens that were displayed.

When we call `popToRoot` and `push` together, the root screen won't be displayed, we will see only the pushed screen. So, the callstack won't be popped to 1 and increased to 2 like the screens stack. It just will be increased to 4 (or 5).

That's why we need a `PopToRootHack`. It works like a screen. Just `screenManager.push(new PopToRootHack)`.
How it works inside?
It will popToRoot and wait until the main thread is free (that means the root screen is already displayed) and then it will push our desired screen. So we will have screens stack and callstack in a consistent state.

I hope you understood something.

Reply:
You can see this callstack in the bottom right corner on the demo. It's a `steps` section.

Reply:
Я в этом вашем Android Auto уже настолько преисполнился, что я как будто бы уже сто триллионов миллиардов лет проживаю на триллионах и триллионах таких же планет, как эта Земля, мне этот мир абсолютно понятен, и я здесь ищу только одного - покоя, умиротворения и вот этой гармонии, от слияния с бесконечно вечным, от созерцания великого фрактального подобия и от вот этого замечательного всеединства существа, бесконечно вечного, куда ни посмотри, хоть вглубь - бесконечно малое, хоть ввысь - бесконечное большое, понимаешь?

![image](https://github.com/organicmaps/organicmaps/assets/10351358/7ddf4383-3940-4892-a508-adf791c5c791)

Reply:
With max 5 templates in a stack the whole UX should be simplified and done differently. 

Is it possible to make a (simplified) diagram of the whole UX, including screens and user actions? Can we make it as simple as possible to get a release ASAP, without repeating the main phone app features? Will ot work on non-touch devices?

Reply:
> With max 5 templates in a stack the whole UX should be simplified and done differently.

I gave you an example of the workflow with search. Any ideas on how to make it simple? 

> Is it possible to make a (simplified) diagram of the whole UX, including screens and user actions?

Yes, it's possible. But I don't see any sense to spend time on it. It won't help us to decrease the number of templates in the stack.

> Can we make it as simple as possible to get a release ASAP

From my pov, it's already as simple as possible. We support only the features that are required during the navigation with AA. I'm pretty sure that simplification (in our case) will harm the UX.

> Will ot work on non-touch devices?

Again, yes. Everything works both on touch and non-touch devices.

Btw, Yandex.Maps uses the same approach.

Reply:
A simpler UX can be done like it is already done in CarPlay:
1. Map screen. User presses "Destination" button.
2. Recent destinations screen with Bookmarks and Search buttons. User presses Search.
3. Search screen with a search box that shows results as you type. When any result from the list is selected, user returns back to the map screen, but with additional control, like this: <img width="398" alt="image" src="https://github.com/organicmaps/organicmaps/assets/170263/aa666926-0f72-4a52-a05a-1d256a0bb915">
4. User can select different search results with a joystick or by clicking left and right arrows. And see the route preview on the map. Back button shows a clean map again.